File: ruby/all-your-base/all_your_base.rb
class BaseConverter
  attr_reader :input_base, :digits, :output_base

  def self.convert(input_base, digits, output_base)

    new(input_base, digits, output_base).convert

    #raise ArgumentError unless valid_input?
    #raise ArgumentError if input_base <= 1 || output_base <= 1 || (digits != [] && digits.min < 0) || (digits != [] && digits.max >= input_base)

    #number = digits.reduce(0) {|product, d| product * input_base + d}

    # converted_number = []
    # if digits.reduce(:+) == 0
    #   converted_number = [0]
    # else
    #   while number > 0
    #     converted_number << number % output_base
    #     number = number / output_base
    #   end
    #   converted_number = converted_number.reverse
    # end
    # converted_number
  end

  def initialize(input_base, digits, output_base)
    @input_base, @digits, @output_base = input_base, digits, output_base

    #raise ArgumentError unless valid_input?
  end

  def valid_input?
    input_base <= 1 || output_base <= 1 || (digits != [] && digits.min < 0) || (digits != [] && digits.max >= input_base)
  end

  def numbers
    #raise ArgumentError unless valid_input?

    digits.reduce(0) {|product, d| product * input_base + d}
  end

  def convert
    raise ArgumentError unless valid_input?

    converted_number = []
    if digits.reduce(:+) == 0
      converted_number = [0]
    else
      while numbers > 0
        converted_number << numbers % output_base
        numbers = numbers / output_base
      end
      converted_number = converted_number.reverse
    end
    converted_number
  end
end

module BookKeeping
  VERSION = 2
end

File: JavaScript/all-your-base/all-your-base.js
export function convert(digits, input_base, output_base) {
    if (!isValid(input_base)) {
        throw new Error("Wrong input base") 
    }

    if (!isValid(output_base)) {
        throw new Error("Wrong output base") 
    }

    if (!isInputValid(digits, input_base)) {
        throw new Error("Input has wrong format") 
    }
     
    let number = digits.reduce((product, d) => product * input_base + d, 0)
    let converted_number = number === 0 ? [0] : []

    while (number > 0) {
      converted_number.push(number % output_base)
      number = Math.floor(number / output_base)
    }

    return converted_number.reverse()
}

function isValid(base) {
    return base > 1 && Number.isInteger(base)
}

function isInputValid(digits, base) {
    return (digits.length > 0 && Math.min(...digits) >= 0 && Math.max(...digits) < base) && (digits[0] !== 0 || digits.length === 1) 
}